server/src/main/java/org/apache/iotdb/db/conf/adapter/ActiveTimeSeriesCounter.java

```diff
@@ -0,0 +1,140 @@
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ActiveTimeSeriesCounter.class);
+  /**
+   * Map[StorageGroup, HyperLogLogCounter]
+   */
+  private static Map<String, HyperLogLog> storageGroupHllMap = new ConcurrentHashMap<>();
+
+  /**
+   * Map[StorageGroup, ActiveTimeSeriesRatio]
+   */
+  private static Map<String, Double> activeRatioMap = new ConcurrentHashMap<>();
+
+  /**
+   * Map[StorageGroup, ActiveTimeSeriesNumber]
+   */
+  private static Map<String, Long> activeTimeSeriesNumMap = new ConcurrentHashMap<>();
+
+  /**
+   * LOG2M decide the precision of the HyperLogLog algorithm
+   */
+  private static final int LOG2M = 13;
+
+  private ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+
+  @Override
+  public void init(String storageGroup) {
+    storageGroupHllMap.put(storageGroup, new HyperLogLog(LOG2M));
+    activeRatioMap.put(storageGroup, 0D);
+    activeTimeSeriesNumMap.put(storageGroup, 0L);
+  }
+
+  @Override
+  public void offer(String storageGroup, String device, String measurement) {
+    try {
+      storageGroupHllMap.get(storageGroup).offer(device + measurement);
+    } catch (Exception e) {
+      storageGroupHllMap.putIfAbsent(storageGroup, new HyperLogLog(LOG2M));
+      storageGroupHllMap.get(storageGroup).offer(device + measurement);
+      LOGGER.error("Register active time series root.{}.{}.{} failed", storageGroup, device,
+          measurement, e);
+    }
+  }
+
+  @Override
+  public void updateActiveRatio(String storageGroup) {
+    lock.writeLock().lock();
+    try {
+      // update the active time series number in the newest memtable to be flushed
+      activeTimeSeriesNumMap.put(storageGroup, storageGroupHllMap.get(storageGroup).cardinality());
+      // initialize the HLL counter
+      storageGroupHllMap.put(storageGroup, new HyperLogLog(LOG2M));
+
+      double totalActiveTsNum = 0;
+      LOGGER.debug("{}: updating active ratio", Thread.currentThread().getName());
+      for (double number : activeTimeSeriesNumMap.values()) {
+        totalActiveTsNum += number;
+      }
+      for (Map.Entry<String, Long> entry : activeTimeSeriesNumMap.entrySet()) {
+        double activeRatio = 0;
+        if (totalActiveTsNum > 0) {
+          activeRatio = entry.getValue() / totalActiveTsNum;
+        }
+        activeRatioMap.put(entry.getKey(), activeRatio);
+        LOGGER.debug("{}: storage group {} has active ratio {}", Thread.currentThread().getName(),
+            entry.getKey(), activeRatio);
+      }
+    } catch (Exception e) {
+      LOGGER.error("Update {} active ratio failed", storageGroup, e);
+    } finally {
+      lock.writeLock().unlock();
+    }
+  }
+
+  @Override
+  public double getActiveRatio(String storageGroup) {
+    lock.writeLock().lock();
+    double ratio;
+    try {
+      ratio = activeRatioMap.get(storageGroup);
+    } catch (Exception e) {
+      LOGGER.error("Get {} active ratio failed", storageGroup, e);
```

Comment:
Why do you try catch here? Is there any chance to occur an exception?

Reply:
Just in case some unexpected test class exceptions, though I have fixed them for now.

Reply:
I think there is a case that when  the memtable is going to be flushed after the storage group is deleted, it may cause the exception. And for any exception situations we may not aware currently or in the future, it is necessary to try catch here.